text/src/autogluon/text/automm/predictor.py

```diff
@@ -433,67 +421,130 @@ def fit(
+                trainable=self._hpo_fit_wrapper,
+                trainable_args=_fit_args,
+                search_space=search_space,
+                hyperparameter_tune_kwargs=hyperparameter_tune_kwargs,
+                metric=metric,
+                mode=mode,
+                save_dir=save_path,
+                ray_tune_adapter=AutommRayTuneAdapter(),
+                total_resources=resources,
+                time_budget_s=time_budget_s,
+                keep_checkpoints_num=3,  # TODO: find a way to extract this from config. Might need to separate generate config and trial specific config
+                checkpoint_score_attr=metric,
+                verbose=2
+            )
+        except EmptySearchSpace:
+            raise ValueError("Please provide a search space in order to do hyperparameter tune")
+        except Exception as e:
+            raise e
+        else:
+            # find the best trial
+            best_trial = analysis.get_best_trial(
+                metric=metric,
+                mode=mode,
+            )
+            if best_trial is None:
+                raise ValueError("AutoMMPredictor wasn't able to find the best trial. It's likely that the time is not enough to train a single epoch for trials.")
+            # clean up other trials
+            logger.info('Removing non-optimal trials and only keep the best one.')
+            cleanup_trials(save_path, best_trial.trial_id)
+            best_trial_path = os.path.join(save_path, best_trial.trial_id)
+            # reload the predictor metadata
+            predictor = self.__class__._load_metadata(path=best_trial_path)
+            predictor._save_path = best_trial_path
+            # construct the model
+            model = create_model(
+                config=predictor._config,
+                num_classes=predictor._output_shape,
+                num_numerical_columns=len(predictor._df_preprocessor.numerical_feature_names),
+                num_categories=predictor._df_preprocessor.categorical_num_categories,
+                pretrained=False # set "pretrain=False" to prevent downloading online models
+            )
+            predictor._model = model
+            # average checkpoint
+            top_k_model_paths = [
+                                    os.path.relpath(os.path.join(path, RAY_TUNE_CHECKPOINT))
+                                    for path, _ in
+                                    sorted(
+                                            list(analysis.get_trial_checkpoints_paths(best_trial, metric=metric)),
+                                            key=lambda ele: ele[1],
+                                            reverse=(mode == MAX)
+                                        )
+                                ]
+            last_ckpt_path = analysis.get_last_checkpoint(best_trial)
+            predictor._top_k_average(
+                model=predictor._model,
+                save_path=predictor._save_path,
```

Comment:
Do we need to use `save_path` or `predictor._save_path`?

Reply:
Should be `predictor._save_path`, it is the actual trial saving path, which is one level deeper than `save_path`. For example, if the user specified `save_path` as `AutommModels` and the best trial is `trial_abc`, then the `predictor._save_path` would be `AutommModels/trial_abc`

Reply:
The issue here is that while performing HPO, we train multiple predictors/trials. And we save them under `save_path/trial_id`; therefore, the resulting artifacts will be in the trial folder. We could move everything one level up in the very end but it could be expensive given that the checkpoint could be large

Reply:
Moving large files could be expensive. Think about you moving files at gigabytes level on your laptop, it could take a while. I'm not sure how big could the artifact be, but we can do it for now, and maybe consider alternatives if this hurts our performance a lot.